### ISSUE TITLE
Significance of micro_batch_size  and global_batch_size in the pre-training scripts

### ISSUE THREAD

**Issue by manideep2510** [Lightning-AI/litgpt#1116]
Hi

I've been playing around with the pre-training scripts (`pretrain/redpajama.py`) and found these two parameters (`micro_batch_size`  and `global_batch_size`) a bit confusing.

From what I could understand, `micro_batch_size` is useful to set gradient accumulation in fine-tuning scripts, and I could see a code block for this in `args.py` where `gradient_accumulation_iters()` is using `micro_batch_size` to set `gradient_accumulation_iters`. 

However, in the redpajama pre-training script, `micro_batch_size` is directly being used in the dataloader as the batch size, and `global_batch_size` is never utilized in the script. In addition, changing `global_batch_size` does not have any effect on GPU consumption or the `iter time` with the pre-training script. Only `micro_batch_size` shows an effect on the GPU utilization, as a normal batch size would.

Could someone kindly help me understand the significance of both the `micro_batch_size`  and `global_batch_size` parameters in the pre-training scripts, on which I should be focussing on to change the batch size and the behavior I should expect?

Thanks!

**Comment by rasbt** [comment]
> From what I could understand, micro_batch_size is useful to set gradient accumulation in fine-tuning scripts, and I could see a code block for this in args.py where gradient_accumulation_iters() is using micro_batch_size to set gradient_accumulation_iters.


In general, that's true. I.e., if you have a global batch size of 8 and a micro batch size of 4, then 

1. the code will load a batch of size 4, do the forward & backward steps
2. load another batch of size 4, do the forward & backward steps
3. apply the weight update & reset the gradients

So, you effectively have a gradient accumulation there. 

In contrast, if global batch size == micro batch size, then you don't have gradient accumulation (just the regular gradient updates).

> Only micro_batch_size shows an effect on the GPU utilization, as a normal batch size would.

It's expected that the global batch size does not affect the memory. Only the micro batch size does. That's because of the gradient accumulation described above. I.e., if you can't train a model with batch size 8, you can set the global batch size of 8 and the micro batch size to a smaller value to get the same loss and updates that you would get when training with the larger batch size, but you process the data in smaller chunks (the micro batch size).

Regarding why the global batch size is not used in the code ... I think it looks like that because it's a bit refactored. I.e., it gets used to calculate the gradient accumulation steps here: https://github.com/Lightning-AI/litgpt/blob/f241d94df59d82b2017bfdcd3800ac8779eb45f5/lit_gpt/args.py#L44

For instance, if you have a global batch size of 8 and a micro batch size of 4, then that's 8 / 4 = 2 gradient accumulation steps 

I hope this makes sense. Please don't hesitate to reach out with follow up questions.

**Comment by manideep2510** [comment]
> For instance, if you have a global batch size of 8 and a micro batch size of 4, then that's 8 / 4 = 2 gradient accumulation steps

This makes more sense now. Thank you @rasbt!


> Could you perhaps repeat the experiment with a micro batch size of 1 and a global batch size as large as possible as your GPU supports?

Here are some stats on 2 x A6000(48GB) system with `tiny-llama-1.1b` pre-training on the RedPajama sample dataset and with a training precision config of `bf16-true`.

| micro_batch_size    | global_batch_size | GPU Consumption (MiB) | iter time (~ ms)|
| -------- | ------- |-------- | ------- |
| 1  | 10   | 2 x 9288 | 490 |
| 1 | 100     | 2 x 9288 | 490 |
| 1    | 1000    | 2 x 9288 | 480 |
| 10    | 100   | 2 x 20782 | 2660 |
| 10   | 1000   | 2 x 20782 | 2660 |

But maybe this makes sense. Because, as the `global_batch_size` increases, the frequency of `optimizer.step` decreases, i.e., it is only run after each `gradient_accumulation_iters` number of steps, which is directly proportional to `global_batch_size` inducing gradient accumulation. This, however, would not increase the memory consumption because we are not saving copies of the gradients separately across iterations but rather with the model's gradients themselves (from what I could understand), which accumulate if not zeroed out after each iteration. So, `global_batch_size` should technically not increase the GPU consumption or the iter time, and only `micro_batch_size` should, as it directly increases the samples per batch that the model processes. I think a larger `global_batch_size` will result in a larger `gradient_accumulation_iters`, making `optimizer.step`  less frequent, decreasing the overall training time (slightly), as we would save some time by not running `optimizer.step` for all steps.

**Comment by rasbt** [comment]
Glad this helped, and your reasoning makes sense. I'd add that while the iter time seems to stay constant here, the time until you reach training convergence will like increase by a lot because the model gets less frequently updated. But yeah, it's always a balance between less-noisy updates (large batch size), preventing to get stuck in local minima (some noise is good), and updating the model frequently (small batch size).

**Comment by manideep2510** [comment]
Yeah, makes sense. The convergence time would be higher with gradient accumulation.

**Comment by gagangayari** [comment]
> > From what I could understand, micro_batch_size is useful to set gradient accumulation in fine-tuning scripts, and I could see a code block for this in args.py where gradient_accumulation_iters() is using micro_batch_size to set gradient_accumulation_iters.
> 
> In general, that's true. I.e., if you have a global batch size of 8 and a micro batch size of 4, then
> 
> 1. the code will load a batch of size 4, do the forward & backward steps
> 2. load another batch of size 4, do the forward & backward steps
> 3. apply the weight update & reset the gradients
> 
> So, you effectively have a gradient accumulation there.
> 
> In contrast, if global batch size == micro batch size, then you don't have gradient accumulation (just the regular gradient updates).
> 
> > Only micro_batch_size shows an effect on the GPU utilization, as a normal batch size would.
> 
> It's expected that the global batch size does not affect the memory. Only the micro batch size does. That's because of the gradient accumulation described above. I.e., if you can't train a model with batch size 8, you can set the global batch size of 8 and the micro batch size to a smaller value to get the same loss and updates that you would get when training with the larger batch size, but you process the data in smaller chunks (the micro batch size).
> 
> Regarding why the global batch size is not used in the code ... I think it looks like that because it's a bit refactored. I.e., it gets used to calculate the gradient accumulation steps here:
> 
> [litgpt/lit_gpt/args.py](https://github.com/Lightning-AI/litgpt/blob/f241d94df59d82b2017bfdcd3800ac8779eb45f5/lit_gpt/args.py#L44)
> 
> Line 44 in [f241d94](/Lightning-AI/litgpt/commit/f241d94df59d82b2017bfdcd3800ac8779eb45f5)
> 
>  def gradient_accumulation_iters(self, devices: int) -> int: 
> For instance, if you have a global batch size of 8 and a micro batch size of 4, then that's 8 / 4 = 2 gradient accumulation steps
> 
> I hope this makes sense. Please don't hesitate to reach out with follow up questions.

Might be a silly one. But does this count as a single step or 2 steps (when you have micro batch size =4 and global batch size= 8).
